### ISSUE TITLE
Time calculation is off

### ISSUE THREAD

**Issue by nicolas-chaulet** [connormanning/arbiter#23]
Hi,
I was using this code from entwine and realized that the time calculation coming from the `asUnix` method was wrong. It is 1 hour off due to the difference between summer and winter time (I am in the UK). I managed to nail down where it goes wrong but I am not sure what would be the best way to fix that. 

Basically, the `Time(std::string)` constructor returns `-3600` instead of `0` for "1970-01-01T00:00:00Z".
Any suggestion about how to fix that?
Thanks a lot!

**Comment by connormanning** [comment]
I think I'd need to be able to reproduce this to be able to tell.  Is this on Windows or Unix?  I added a test for this at d1c65d2, which passes for me.  If you can give steps to produce an environment, or code that sets the timezone, such that this test fails for me, then I could debug.

**Comment by nicolas-chaulet** [comment]
Yes, not that simple, test actually pass just fine, it is really related to the call to `std::time(nullptr)` which seems to return a time that does not need correction for me... I am on Unix, let me see if I can figure out a simple test that would break it. 

I have added the following to the test, which make them fail:
```
    Time epoch("1970-01-01T00:00:00Z");
    EXPECT_EQ(epoch.raw(),0);
```

where 
```
int64_t Time::raw() const {
  return m_time;
}
```
But maybe it makes sense that `m_time` is not 0. Not sure.

**Comment by connormanning** [comment]
> the time calculation coming from the asUnix method was wrong

In what way is the `asUnix` method result incorrect?  I would think the subtraction of the epoch would correct for this.

> But maybe it makes sense that m_time is not 0. Not sure.

I'm also not positive about this, but I could see how it could be the correct behavior.

Since the private member `m_time` is not exposed it sounds like this is an issue with an implementation detail - can you elaborate on the use-case where the issue occurs?

**Comment by nicolas-chaulet** [comment]
The test below breaks it for me actually:
```
Time now("2019-01-03T17:39:11Z");
EXPECT_GT(now.asUnix()-1546537151,0);
EXPECT_LT(now.asUnix()-1546537151,1);
```
reference timestamp was taken from there:
https://www.unixtimestamp.com/index.php

**Comment by connormanning** [comment]
I'm confused at that test: wouldn't you expect the value of `now.asUnix() - 1546537151` to be zero?  Why is it testing for `> 0` and `< 1`?

**Comment by nicolas-chaulet** [comment]
Yes, sorry, I edited the comment...

**Comment by connormanning** [comment]
I've reproduced the issue with your test using `TZ=Europe/London ./test/arbiter-test` from my build directory.  I don't have a fix yet.  If you'd like to debug as well, you should be able to reproduce my behavior (with which the test passes) for comparison with `TZ=America/Chicago ./test/arbiter-test` on Unix.  I'll look into this further in a few hours.

**Comment by nicolas-chaulet** [comment]
Yes, I managed to narrow it down. This is due to daylight saving issues, daylight saving was introduced in 1972 which explains why the test on epoch works fine but not the test on dates after 1972. I think the source of the problem is the calculation of the `utcOffset` which should take those savings into account but at the same time `std::mktime` seems to include some logic for local correction as well. Thanks for your help!

**Comment by connormanning** [comment]
Can you take a look at the `time` branch and see if this fixes your issue?  See [here](https://github.com/connormanning/arbiter/commit/ade5f744e30ee9b416fb1b22f6a66748835f5eb9) for the potential fix.

**Comment by connormanning** [comment]
I believe the issue was that the local->UTC offset was always calculated from the current time - where if you pass in a time string, then that offset should be calculated relative to that time instead, due to potential DST scenarios.

**Comment by nicolas-chaulet** [comment]
Works for me! Thank you so much @connormanning!
